File: mcp/mcp_server.py
import asyncio
import requests

from modelcontextprotocol.server import Server
from modelcontextprotocol.server.stdio import stdio_server
from modelcontextprotocol.types import TextContent

server = Server(
    name="codex-refactor-mcp",
    version="1.0.0",
)

def call_backend(path, body):
    url = f"http://localhost:4000{path}"
    res = requests.post(url, json=body)
    return res.json()

@server.call_tool()
async def analyze_code(tool_name: str, args: dict):
    code = args.get("code")
    filename = args.get("filename")
    resp = call_backend("/analyze", {"code": code, "filename": filename})
    return [TextContent(type="text", text=str(resp))]

@server.call_tool()
async def suggest_refactors(tool_name: str, args: dict):
    code = args.get("code")
    filename = args.get("filename")
    resp = call_backend("/suggest", {"code": code, "filename": filename})
    return [TextContent(type="text", text=str(resp))]

async def main():
    print("🔥 Codex Refactor MCP running...", flush=True)
    async with stdio_server() as (read_stream, write_stream):
        await server.run(
            read_stream,
            write_stream,
            server.create_initialization_options(),
        )

if __name__ == "__main__":
    asyncio.run(main())

File: codex_mcp/mcp-server.js
// mcp/mcp-server.js
// FINAL — 100% working with @modelcontextprotocol/sdk@1.23.0

import {
  Server,
  NodeStdioServerTransport
} from "@modelcontextprotocol/sdk/server";

// ----------------------------------------
// Backend Proxy Helper
// ----------------------------------------
async function callBackend(path, body) {
  const url = `http://localhost:4000${path}`;

  const res = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });

  return res.json();
}

// ----------------------------------------
// CREATE MCP SERVER
// ----------------------------------------
const server = new Server(
  {
    name: "codex-refactor-mcp",
    version: "1.0.0",
  },
  {
    tools: {
      // ----------------------------------------
      // Tool 1 — Analyze Code
      // ----------------------------------------
      analyze_code: {
        description: "Analyze source code AST and structure.",
        inputSchema: {
          type: "object",
          properties: {
            code: { type: "string" }
          },
          required: ["code"],
        },
        handler: async ({ code }) => {
          try {
            const resp = await callBackend("/analyze", { code });
            return { ok: true, analysis: resp.analysis };
          } catch (err) {
            return { ok: false, error: err.message };
          }
        }
      },

      // ----------------------------------------
      // Tool 2 — Suggest Refactors
      // ----------------------------------------
      suggest_refactors: {
        description:
          "Suggest AST-based refactors and return full patched file output.",
        inputSchema: {
          type: "object",
          properties: {
            code: { type: "string" }
          },
          required: ["code"],
        },
        handler: async ({ code }) => {
          try {
            const resp = await callBackend("/suggest", { code });
            return { ok: true, suggestions: resp.suggestions };
          } catch (err) {
            return { ok: false, error: err.message };
          }
        }
      }
    }
  }
);

// ----------------------------------------
// START SERVER USING STDIO TRANSPORT
// ----------------------------------------
const transport = new NodeStdioServerTransport();

server.start(transport);

console.error("🚀 codex-refactor-mcp server started (NodeStdioServerTransport)");
